File: redcell_classifier.py
"""Optional second-stage scorer — the Python half of the pair. See redcell_classifier.js.

Kept byte-for-byte equivalent to the JS scorer and asserted so by the parity test. The two must
agree: the Worker serves the JS engine and every published measurement is produced by the Python
one, so a divergence would mean the numbers on the site describe a different product than the
one customers call.
"""
import json
import math
import os
import re
from collections import Counter

_MODEL_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), "models", "injection_lr.json")
_TOKEN = re.compile(r"[^\W\d_]+", re.UNICODE)
_MAX_CHARS = 4000

with open(_MODEL_PATH, encoding="utf-8") as _f:
    _M = json.load(_f)
_W = {int(k): v for k, v in _M["w"].items()}
_B = _M["b"]
_BUCKETS = _M["buckets"]
THRESHOLD = _M["thr"]


def _fnv1a(s):
    """32-bit FNV-1a. Python's built-in hash() is randomised per process — weights trained in one
    run scored ~0.11 on everything in the next — and it cannot be reproduced in JS."""
    h = 0x811C9DC5
    for ch in s:
        h ^= ord(ch) & 0xFF
        h = (h * 0x01000193) & 0xFFFFFFFF
    return h


def score(text):
    """Probability in [0,1] that the text is an injection attempt."""
    if not isinstance(text, str) or not text:
        return 0.0
    t = text.lower()[:_MAX_CHARS]
    words = _TOKEN.findall(t)
    f = Counter()
    for w in words:
        f[_fnv1a("u:" + w) % _BUCKETS] += 1
    for a, b in zip(words, words[1:]):
        f[_fnv1a("b:" + a + "_" + b) % _BUCKETS] += 1
    f[_BUCKETS - 1] += min(len(words) / 50.0, 4.0)
    f[_BUCKETS - 2] += t.count("\n") / 5.0
    f[_BUCKETS - 3] += sum(c in "{}[]<>|" for c in t) / 10.0

    norm = math.sqrt(sum(v * v for v in f.values())) or 1.0
    z = _B + sum(_W.get(k, 0.0) * (v / norm) for k, v in f.items())
    return 1 / (1 + math.exp(-max(-30, min(30, z))))

File: redcell_classifier.js
/**
 * Optional second-stage scorer: logistic regression over hashed word uni/bigrams.
 *
 * The rule engine is precise and incomplete — measured at 93-100% precision but 17-28% recall on
 * public third-party corpora. This adds recall on traffic that resembles those corpora, at the
 * cost of a model that has to be trusted rather than read. It is therefore OFF by default and
 * returns a score the caller can act on, rather than silently changing verdicts.
 *
 * No API, no dependency, no network: 3,000 float weights and a dot product. Sub-millisecond.
 *
 * Parity with the Python scorer is not optional — both engines must agree, so the tokenizer uses
 * \p{L} rather than \w. JS's \w is ASCII-only, and the training data is half German; a \w-based
 * split would have silently disagreed with Python on every umlaut.
 */
import MODEL from './redcell_model.js';

const TOKEN = /\p{L}+/gu;
const MAX_CHARS = 4000;

function fnv1a(s) {
  let h = 0x811c9dc5;
  // Iterate CODE POINTS, not UTF-16 units. Python's ord() walks code points, so charCodeAt made
  // the two scorers disagree on any astral character — caught on mathematical bold letters,
  // which are exactly what an obfuscated payload uses.
  for (const ch of s) {
    h ^= ch.codePointAt(0) & 0xff;
    h = Math.imul(h, 0x01000193) >>> 0;
  }
  return h;
}

/** Raw probability in [0,1] that the text is an injection attempt. */
export function score(text) {
  if (typeof text !== 'string' || !text) return 0;
  const t = text.toLowerCase().slice(0, MAX_CHARS);
  const words = t.match(TOKEN) || [];
  const B = MODEL.buckets;
  const f = new Map();
  const bump = (k, d) => f.set(k, (f.get(k) || 0) + d);
  for (const w of words) bump(fnv1a('u:' + w) % B, 1);
  for (let i = 0; i + 1 < words.length; i++) bump(fnv1a('b:' + words[i] + '_' + words[i + 1]) % B, 1);
  bump(B - 1, Math.min(words.length / 50, 4));
  let nl = 0; for (let i = 0; i < t.length; i++) if (t.charCodeAt(i) === 10) nl++;
  bump(B - 2, nl / 5);
  let br = 0; for (const c of t) if ('{}[]<>|'.indexOf(c) >= 0) br++;
  bump(B - 3, br / 10);

  let norm = 0;
  for (const v of f.values()) norm += v * v;
  norm = Math.sqrt(norm) || 1;

  let z = MODEL.b;
  for (const [k, v] of f) {
    const w = MODEL.w[k];
    if (w !== undefined) z += w * (v / norm);
  }
  z = Math.max(-30, Math.min(30, z));
  return 1 / (1 + Math.exp(-z));
}

/** The threshold chosen for zero false positives on ordinary business traffic. */
export const THRESHOLD = MODEL.thr;

export default { score, THRESHOLD };
